api/v1/verticadb_webhook.go

```diff
@@ -1261,6 +1285,57 @@ func checkInt64PtrChange(prefix *field.Path, fieldName string,
 	return allErrs
 }
 
+// checkImmutableSubclusterDuringUpgrade will ensure we don't scale, add or
+// remove subclusters during a replicated upgrade.
+func (v *VerticaDB) checkImmutableSubclusterDuringUpgrade(oldObj *VerticaDB, allErrs field.ErrorList) field.ErrorList {
+	// This entire check can be skipped if we aren't doing replicated upgrade.
+	if oldObj.Status.UpgradeState == nil || len(oldObj.Status.UpgradeState.ReplicaGroups) == 0 {
+		return allErrs
+	}
+
+	// The subclusters defined in status.upgradeState.replicaGroups must stay
+	// the same. We do it this way to allow the operator to add subclusters
+	// during replicated upgrade, but keep all of the other subclusters the
+	// same.
+	oldScMap := oldObj.GenSubclusterMap()
+	newScMap := v.GenSubclusterMap()
+	path := field.NewPath("spec").Child("subclusters")
+	for i := range v.Status.UpgradeState.ReplicaGroups {
+		for _, scName := range v.Status.UpgradeState.ReplicaGroups[i] {
+			oldSc, oldScFound := oldScMap[scName]
+			newSc, newScFound := newScMap[scName]
+			// Cannot remove a subcluster that is in the replica group
+			if !newScFound {
+				err := field.Invalid(path,
+					v.Spec.Subclusters,
+					fmt.Sprintf("Cannot remove subcluster %q while database is being upgraded", scName))
+				allErrs = append(allErrs, err)
+				continue
+			}
+			// The operator can create new subclusters that are used for the
+			// upgrade process. These must be secondary subclusters.
+			if !oldScFound {
```

Comment:
I am confused by this oldScFound/newScFound checks. How can we check them both  together?

Reply:
oldScFound = true, newScFound = false => this can happen if a subcluster is attempted to be removed
oldScFound = false, newScFound = true => this can happen if a new subcluster is added. We allow this as long as the subcluster is a secondary.